### PR TITLE
WFLY-9004 Artemis leaks JGroups channels on reload

### PR DESCRIPTION
**Pull request by rhusar**
Jira
https://issues.jboss.org/browse/WFLY-9004

### CODE REVIEW

**Comment by rhusar** [comment]
@pferraro Please review; this fixes the leak, but let me know if you think we should do this another way..

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ForkChannelBroadcastEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class ForkChannelBroadcastEndpoint extends JGroupsChannelBroadcastEndpoint {
+
+    public ForkChannelBroadcastEndpoint(JChannelManager manager, JChannel jChannel, final String channelName) {
+        super(manager, jChannel, channelName);
+    }
+
+    @Override
+    protected synchronized void internalCloseChannel(JChannelWrapper channel) {
+        channel.close(true);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ForkChannelBroadcastEndpoint.java]: Future-proof this by using JChannel.close().

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java [hunk 2]
```diff
@@ -153,7 +152,7 @@ static DiscoveryGroupConfiguration createDiscoveryGroupConfiguration(final Strin
         final long refreshTimeout = config.getRefreshTimeout();
         final long initialWaitTimeout = config.getDiscoveryInitialWaitTimeout();
 
-        final BroadcastEndpointFactory endpointFactory = new ChannelBroadcastEndpointFactory(channel, channelName);
+        final BroadcastEndpointFactory endpointFactory = new ForkChannelBroadcastEndpointFactory(channel, channelName);
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java]: While the channel is question is a ForkChannel by default, it is not necessarily so.  Thus the name of this class isn't quite appropriate.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ForkChannelBroadcastEndpointFactory.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.activemq.artemis.api.core.jgroups.JChannelManager;
+import org.jgroups.JChannel;
+
+/**
+ * An implementation of {@link org.apache.activemq.artemis.api.core.BroadcastEndpointFactory} which uses a fork channel
+ * and closes the channel when no longer in use.
+ *
+ * @author Radoslav Husar
+ */
+public class ForkChannelBroadcastEndpointFactory extends ChannelBroadcastEndpointFactory {
```
Comment (pferraro) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ForkChannelBroadcastEndpointFactory.java]: See above comment regarding the name.  The logic for closing the channel would be the same regardless of implementation.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ForkChannelBroadcastEndpoint.java (hunk 1 above)
Comment (rhusar) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ForkChannelBroadcastEndpoint.java]: That would not fix the problem; firstly the wrapper uses a reference counter and only closes the channel when last consumer of the channel closes, but mainly we need to trigger org.apache.activemq.artemis.api.core.jgroups.JChannelWrapper#close which removes the reference from its internal structure.

**Pull request merged**